File: tests/netDxf.Conformance/TypedCommentTests.cs
using netDxf;
using netDxf.Header;
using netDxf.IO;

namespace NetDxf.Conformance;

internal static partial class Program
{
    private static void RegisterTypedCommentTests()
    {
        foreach (DxfVersion version in SupportedVersions)
        {
            DxfVersion v = version;
            foreach (string mode in new[] { "every-tag", "runs", "section", "header", "table", "subclass", "geometry", "xdata", "eof", "leading" })
            {
                string m = mode;
                Run($"typed/comments/{v}/{m}", () => TypedCommentRoundTrip(v, m));
            }
            Run($"typed/comments/{v}/binary-control", () => TypedCommentBinaryControl(v));
            foreach (bool comment in new[] { false, true })
            {
                bool c = comment;
                Run($"typed/comments/{v}/malformed/{c}", () => TypedCommentMalformed(v, c));
            }
        }
    }

    private static List<DxfTag> TypedCommentTags(DxfVersion version, string mode)
    {
        var original = HatchPatternValidationTags(version);
        var result = new List<DxfTag> { new(999, "leading source comment") };
        string section = "";
        for (int i = 0; i < original.Count; i++)
        {
            var tag = original[i];
            if (tag.Code == 2 && i > 0 && original[i - 1].Code == 0 && Equals(original[i - 1].Value, "SECTION")) section = (string)tag.Value;
            bool inject = mode is "every-tag" or "runs" ||
                (mode == "section" && tag.Code is 0 or 2) || (mode == "header" && section == "HEADER") ||
                (mode == "table" && section == "TABLES") || (mode == "subclass" && tag.Code == 100) ||
                (mode == "geometry" && section == "ENTITIES" && tag.Code is >= 10 and < 100) ||
                (mode == "xdata" && tag.Code >= 1000) || (mode == "eof" && tag.Code == 0 && Equals(tag.Value, "EOF"));
            if (inject)
                for (int n = 0; n < (mode == "runs" ? 33 : 1); n++)
                    result.Add(new(999, "0 SECTION 2 HEADER 0 EOF are comment data " + n));
            result.Add(tag);
        }
        return result;
    }

    private static void TypedCommentRoundTrip(DxfVersion version, string mode)
    {
        var tags = TypedCommentTags(version, mode);
        string[] leading = tags.TakeWhile(t => t.Code == 999).Select(t => (string)t.Value).ToArray();
        byte[] bytes = RawFixtureBytes(tags, false);
        using var input = new MemoryStream(bytes);
        Equal(version, DxfDocument.CheckDxfFileVersion(input, out bool binary), "Commented version probe"); Check(!binary, "Text detected as binary.");
        input.Position = 0;
        var doc = DxfDocument.Load(input) ?? throw new InvalidOperationException("Valid commented drawing rejected.");
        Check(doc.Comments.SequenceEqual(leading), "Leading comments changed or interstitial comments leaked into header.");
        HatchPatternCheck(doc, "canonical", 1);
        Equal(new Vector3(20, 30, 40), doc.Entities.Lines.Single().StartPoint, "Comment shifted point fields");
        foreach (bool format in new[] { false, true })
        {
            using var output = new MemoryStream(); Check(doc.Save(output, format), "Commented drawing export failed."); output.Position = 0;
            var raw = DxfRawDocument.Load(output);
            string[] emitted = raw.Tags.Where(t => t.Code == 999).Select(t => (string)t.Value).ToArray();
            Check(emitted.SequenceEqual(format ? Array.Empty<string>() : leading), "Typed comment preservation policy changed.");
            output.Position = 0; var loaded = DxfDocument.Load(output) ?? throw new InvalidOperationException("Commented drawing reload failed.");
            HatchPatternCheck(loaded, "canonical", 1);
            Equal(new Vector3(20, 30, 40), loaded.Entities.Lines.Single().StartPoint, "Following entity changed");
        }
        input.Position = 0;
        var untouched = DxfRawDocument.Load(input);
        using var rawOutput = new MemoryStream(); untouched.Save(rawOutput);
        Check(bytes.SequenceEqual(rawOutput.ToArray()), "Typed comment fix changed exact raw preservation.");
        if (mode == "every-tag") File.WriteAllBytes(Path.Combine(ArtifactDirectory, $"typed-comments-{version}.dxf"), bytes);
        Check(input.CanRead, "Comment wrapper closed caller stream.");
    }

    private static void TypedCommentBinaryControl(DxfVersion version)
    {
        using var input = new MemoryStream(RawFixtureBytes(HatchPatternValidationTags(version), true));
        var doc = DxfDocument.Load(input) ?? throw new InvalidOperationException("Binary behavior changed.");
        HatchPatternCheck(doc, "canonical", 1); Equal(0, doc.Comments.Count, "Binary invented comments");
    }

    private static void TypedCommentMalformed(DxfVersion version, bool comment)
    {
        var tags = TypedCommentTags(version, comment ? "every-tag" : "leading");
        int count = tags.FindIndex(t => t.Code == 78); tags[count] = new(78, (short)-1);
        using var input = new MemoryStream(RawFixtureBytes(tags, false));
#if DEBUG
        try { DxfDocument.Load(input); throw new InvalidOperationException("Comments hid a malformed HATCH count."); }
        catch (InvalidDataException error) { Check(error.Message.Contains("HATCH", StringComparison.Ordinal), "Comment filtering lost the underlying diagnostic."); }
#else
        Check(DxfDocument.Load(input) == null, "Comments hid malformed input.");
#endif
    }
}

File: javascript/tests/netDxf.Conformance/TypedCommentTests.js
// Complete original TypedCommentTests.cs. Comment text must never become semantic tags.
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { DxfDocument, DxfRawDocument, DxfTag, Vector3, MemoryStream } from '../../index.js';
import { InvalidDataException, InvalidOperationException } from '../../runtime/Errors.js';
import { GetTypedIOConfiguration } from '../../runtime/TypedDocumentIO.js';
import { HatchPatternValidationTags, HatchPatternCheck } from './HatchPatternValidationTests.js';
import { RawFixtureBytes } from './RawDocumentTests.js';
import { Run, Check, Equal, SupportedVersions, VersionName, BooleanName } from './TestHarness.js';
const artifactDirectory=path.resolve(path.dirname(fileURLToPath(import.meta.url)),'../../artifacts/conformance/fixtures');
const single=items=>{const rows=Array.from(items);Equal(1,rows.length,'Expected one item');return rows[0];};
export function RegisterTypedCommentTests(){for(const v of SupportedVersions){for(const mode of ['every-tag','runs','section','header','table','subclass','geometry','xdata','eof','leading'])Run(`typed/comments/${VersionName(v)}/${mode}`,()=>TypedCommentRoundTrip(v,mode));Run(`typed/comments/${VersionName(v)}/binary-control`,()=>TypedCommentBinaryControl(v));for(const c of [false,true])Run(`typed/comments/${VersionName(v)}/malformed/${BooleanName(c)}`,()=>TypedCommentMalformed(v,c));}}
export function TypedCommentTags(version,mode){const original=HatchPatternValidationTags(version),result=[new DxfTag(999,'leading source comment')];let section='';
  for(let i=0;i<original.length;i++){const tag=original[i];if(tag.Code===2&&i>0&&original[i-1].Code===0&&original[i-1].Value==='SECTION')section=tag.Value;
    const inject=mode==='every-tag'||mode==='runs'||mode==='section'&&(tag.Code===0||tag.Code===2)||mode==='header'&&section==='HEADER'||mode==='table'&&section==='TABLES'||mode==='subclass'&&tag.Code===100||mode==='geometry'&&section==='ENTITIES'&&tag.Code>=10&&tag.Code<100||mode==='xdata'&&tag.Code>=1000||mode==='eof'&&tag.Code===0&&tag.Value==='EOF';
    if(inject)for(let n=0;n<(mode==='runs'?33:1);n++)result.push(new DxfTag(999,'0 SECTION 2 HEADER 0 EOF are comment data '+n));result.push(tag);
  }return result;
}
export function TypedCommentRoundTrip(version,mode){const tags=TypedCommentTags(version,mode),leading=[];for(const tag of tags){if(tag.Code!==999)break;leading.push(tag.Value);}const bytes=RawFixtureBytes(tags,false),input=new MemoryStream(bytes);try{
  const binary={};Equal(version,DxfDocument.CheckDxfFileVersion(input,binary),'Commented version probe');Check(!binary.value,'Text detected as binary.');input.Position=0;const doc=DxfDocument.Load(input);if(!doc)throw new InvalidOperationException('Valid commented drawing rejected.');Equal(leading,Array.from(doc.Comments),'Leading comments changed or interstitial comments leaked into header.');HatchPatternCheck(doc,'canonical',1);Check(new Vector3(20,30,40).Equals(single(doc.Entities.Lines).StartPoint),'Comment shifted point fields');
  for(const format of [false,true]){const output=new MemoryStream();try{Check(doc.Save(output,format),'Commented drawing export failed.');output.Position=0;const raw=DxfRawDocument.Load(output),emitted=Array.from(raw.Tags).filter(t=>t.Code===999).map(t=>t.Value);Equal(format?[]:leading,emitted,'Typed comment preservation policy changed.');output.Position=0;const loaded=DxfDocument.Load(output);if(!loaded)throw new InvalidOperationException('Commented drawing reload failed.');HatchPatternCheck(loaded,'canonical',1);Check(new Vector3(20,30,40).Equals(single(loaded.Entities.Lines).StartPoint),'Following entity changed');}finally{output.Dispose();}}
  input.Position=0;const untouched=DxfRawDocument.Load(input),rawOutput=new MemoryStream();try{untouched.Save(rawOutput);Equal(bytes,rawOutput.ToArray(),'Typed comment fix changed exact raw preservation.');}finally{rawOutput.Dispose();}
  if(mode==='every-tag'){fs.mkdirSync(artifactDirectory,{recursive:true});fs.writeFileSync(path.join(artifactDirectory,`typed-comments-${VersionName(version)}.dxf`),bytes);}Check(input.CanRead,'Comment wrapper closed caller stream.');
}finally{input.Dispose();}}
export function TypedCommentBinaryControl(version){const input=new MemoryStream(RawFixtureBytes(HatchPatternValidationTags(version),true));try{const doc=DxfDocument.Load(input);if(!doc)throw new InvalidOperationException('Binary behavior changed.');HatchPatternCheck(doc,'canonical',1);Equal(0,doc.Comments.Count,'Binary invented comments');}finally{input.Dispose();}}
export function TypedCommentMalformed(version,comment){const tags=TypedCommentTags(version,comment?'every-tag':'leading');tags[tags.findIndex(t=>t.Code===78)]=new DxfTag(78,-1);const input=new MemoryStream(RawFixtureBytes(tags,false));try{
  if(GetTypedIOConfiguration()==='Debug'){try{DxfDocument.Load(input);throw new InvalidOperationException('Comments hid a malformed HATCH count.');}catch(error){Check(error instanceof InvalidDataException,'Expected InvalidDataException');Check(error.message.includes('HATCH'),'Comment filtering lost the underlying diagnostic.');}}else Check(DxfDocument.Load(input)===null,'Comments hid malformed input.');
}finally{input.Dispose();}}
